### PR TITLE
Storyboard: add full type and module information to context

### LINKED ISSUE #404: Support AVPlayerViewController in Storyboard Gen
There currently is an issue with the Storyboard code generation.
When I use an AVPlayerViewController inside the storyboard of an iOS project, SwiftGen defines the type as `UIAvPlayerViewController` which is unknown.

My current workaround: 
```swift
import AVKit

typealias UIAvPlayerViewController = AVPlayerViewController
```

--- 

FYI: The default XML Code from the storyboard file looks like this:
```xml
<avPlayerViewController videoGravity="AVLayerVideoGravityResizeAspect" id="0eg-7u-MRY" sceneMemberID="viewController"/>
```
vs a default `UIViewController`
```xml
<viewController id="DsH-rV-T6N" sceneMemberID="viewController">
```

### PR DESCRIPTION
**Pull request by djbe**
Fixes #404.

`AVPlayerViewController` and `GLKViewController` are special cases that do not need the `UI` prefix. As a long term solution, we're better of moving (parts of) the `className` macro to the parser, so that it can generate the correct class/module information for a scene.

A couple of things to note (some old items from before the refactor of this PR):
- AppKit doesn't have this issue (I've tried all window/view controller types available).
- (OLD) This needs to be handled in the template, as that's where the prefix is added.
- I considered adding a flag to the context if the parser detects a base type that does not need a prefix, but that's an even hackier solution.
- (OLD) Better would be to add the needed prefix in the parser, but I'm not sure how to handle the platform specific prefix (it could be different per storyboard). And it would be a breaking change.
- Long term, we could add a variable to the context with the actual `class` for a scene, that would be the same result as our current (in template) macro `className`.

### CODE REVIEW

**Comment by djbe** [comment]
Damn, didn't think about the module imports... I'm afraid I'll have to implement this:
> Long term, we could add a variable to the context with the actual `class` for a scene, that would be the same result as our current (in template) macro `className`.

File: Tests/Fixtures/Generated/Storyboards-iOS/swift2-context-all-ignore-module.swift [hunk 1]
```diff
@@ -62,43 +64,50 @@ internal enum StoryboardScene {
...
-    internal static let Navigation = SceneType<UINavigationController>(Placeholder.self, identifier: "Navigation")
+    internal static let Navigation = SceneType<UIKit.UINavigationController>(Placeholder.self, identifier: "Navigation")
   }
   internal enum Wizard: StoryboardType {
     internal static let storyboardName = "Wizard"
 
     internal static let initialScene = InitialSceneType<CreateAccViewController>(Wizard.self)
 
-    internal static let AcceptCGU = SceneType<UIViewController>(Wizard.self, identifier: "Accept-CGU")
+    internal static let AcceptCGU = SceneType<UIKit.UIViewController>(Wizard.self, identifier: "Accept-CGU")
```
Comment (AliSoftware) [inline Tests/Fixtures/Generated/Storyboards-iOS/swift2-context-all-ignore-module.swift]: Side note: We should fix that franglicism in the examples one day 😂 (CGU 🇫🇷 = "Conditions Générales d'Utilisation" = TOS 🇺🇸 = "Terms of Service")

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,9 @@
 * Fix memory leak in generated code for Fonts.  
   [Olivier Halligon](https://github.com/AliSoftware)
   [#394](https://github.com/SwiftGen/SwiftGen/issues/394)
+* Storyboards: ensure the templates handle `GLKViewController` and `AVPlayerViewController` correctly.  
+  [David Jennes](https://github.com/djbe) 
+  [#414](https://github.com/SwiftGen/SwiftGen/pull/414)
```
[the remark this answers is not in the document]
Comment (djbe) [reply CHANGELOG.md]: Will do.

File: Sources/SwiftGenKit/Parsers/StoryboardParser.swift [hunk 3]
```diff
@@ -35,37 +56,170 @@ private enum XML {
...
+    }
+
+    private static let tagTypeMap = [
+      "avPlayerViewController": "AVPlayerViewController",
+      "glkViewController": "GLKViewController"
+    ]
+
+    private static let tagModuleMap = [
+      "avPlayerViewController": "AVKit",
+      "glkViewController": "GLKit"
```
[the remark this answers is not in the document]
Comment (djbe) [reply Sources/SwiftGenKit/Parsers/StoryboardParser.swift]: I tried whatever had the words controller and window in it. Most of the rest were views, so they won't affect us in this scenario.

File: Tests/Fixtures/Generated/Storyboards-iOS/swift2-context-all-ignore-module.swift (hunk 1 above)
Comment (djbe) [reply Tests/Fixtures/Generated/Storyboards-iOS/swift2-context-all-ignore-module.swift]: Never even realised that's what it stood for 😆

File: Sources/SwiftGenKit/Parsers/StoryboardParser.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (djbe) [reply Sources/SwiftGenKit/Parsers/StoryboardParser.swift]: I've added a storyboard for iOS and macOS to test all known windows/controllers that can be added using IB. This storyboard should be updated for each new Xcode version. The tagModuleMap should also be re-checked when we (continue to) implement the XIB parser, as we'll need to handle views and other stuff.

**Comment by djbe** [comment]
> error: no type named 'NSPagecontroller' in module 'AppKit'
Seriously IB, wth? 🤦‍♂️

File: Tests/Fixtures/Resources/Storyboards-macOS/Placeholder.storyboard [hunk 4]
```diff
@@ -1,8 +1,7 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.Cocoa.Storyboard.XIB" version="3.0" toolsVersion="10117" systemVersion="15F34" targetRuntime="MacOSX.Cocoa" propertyAccessControl="none" useAutolayout="YES">
+<?xml version="1.0" encoding="UTF-8"?>
+<document type="com.apple.InterfaceBuilder3.Cocoa.Storyboard.XIB" version="3.0" toolsVersion="14109" targetRuntime="MacOSX.Cocoa" propertyAccessControl="none" useAutolayout="YES">
     <dependencies>
-        <deployment identifier="macosx"/>
```
[the remark this answers is not in the document]
Comment (AliSoftware) [reply Tests/Fixtures/Resources/Storyboards-macOS/Placeholder.storyboard]: I mean I see it removed on multiple storyboards it's probably not generated by latest Xcode versions it seams? But just want to be sure we don't depend on it ourselves

[the remark this answers is not in the document]
Comment (djbe) [reply Tests/Fixtures/Resources/Storyboards-macOS/Placeholder.storyboard]: We don't, it got auto-removed when I opened (and re-saved) the file. We use the `targetRuntime="iOS.CocoaTouch"` attribute on the document node.

**Pull request merged**